### PR TITLE
Unpack runtime env to runtime_resource

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

To let workers able to access files other than py codes, we need to unpack the package and add that to working dir.
This PR assumes, there is only one runtime env, which is true for now. Later it needs to be extended to support multiple uris.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/runtime_env.py [hunk 1]
```diff
@@ -340,27 +345,27 @@ def upload_runtime_env_package_if_needed(job_config: JobConfig) -> None:
...
-                logger.debug(
-                    f"{pkg_uri} has existed locally, skip downloading")
-            else:
-                pkg_size = fetch_package(pkg_uri, pkg_file)
-                logger.debug(f"Downloaded {pkg_size} bytes into {pkg_file}")
-        sys.path.insert(0, str(pkg_file))
+            pkg_dir = fetch_package(pkg_uri)
+        sys.path.insert(0, str(pkg_dir))
+    # Right now, multiple pkg_uris are not supported correctly.
+    # We return the last one as working directory
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:370]: We plan to move this to a separate field, right? This seems like it could cause confusion.

File: python/ray/tests/test_runtime_env.py [hunk 2]
```diff
@@ -131,6 +138,24 @@ def test_two_node_module(two_node_cluster, working_dir, client_mode):
...
+    cluster, _ = two_node_cluster
+    (address, env, PKG_DIR) = start_client_server(cluster, client_mode)
+    runtime_env = f"""{{  "working_dir": "{working_dir}" }}"""
+    execute_statement = """
+vals = ray.get([check_file.remote('test_file')] * 1000)
+print(sum([int(v) for v in vals]))
+"""
+    script = driver_script.format(**locals())
+    out = run_string_as_driver(script, env)
+    assert out.strip().split()[-1] == "1000"
```
Comment (edoakes) [inline python/ray/tests/test_runtime_env.py:155]: please add a comment for what this is doing; it's a bit hard to parse for an unfamiliar reader

File: python/ray/_private/runtime_env.py [hunk 3]
```diff
@@ -340,27 +345,27 @@ def upload_runtime_env_package_if_needed(job_config: JobConfig) -> None:
...
 
     Necessary packages required to run the job will be downloaded
     into local file system if it doesn't exist.
 
     Args:
         pkg_uri list(str): Package of the working dir for the runtime env.
-    """
 
+    Return:
+        Working directory is returned.
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:358]: This returns `None` in some cases -- please document that.

File: python/ray/_private/runtime_env.py [hunk 4]
```diff
@@ -340,27 +345,27 @@ def upload_runtime_env_package_if_needed(job_config: JobConfig) -> None:
             logger.info(f"{pkg_uri} has been pushed with {pkg_size} bytes")
 
 
-def ensure_runtime_env_setup(pkg_uris: List[str]) -> None:
+def ensure_runtime_env_setup(pkg_uris: List[str]) -> str:
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:348]:
```suggestion
def ensure_runtime_env_setup(pkg_uris: List[str]) -> Optional[str]:
```

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1244,7 +1244,8 @@ void NodeManager::DeleteLocalURI(const std::string &uri, std::function<void(bool
     cb(true);
   }
 
-  auto from_path = resource_path / boost::filesystem::path(uri.substr(pos + sep.size()));
+  auto from_path =
+      resource_path / boost::filesystem::path(uri.substr(pos + sep.size())).stem();
```
Comment (edoakes) [inline src/ray/raylet/node_manager.cc:1248]: was this a bug?

**Review by edoakes — COMMENTED** [review]
Looks good overall, just some small comments

File: python/ray/_private/runtime_env.py (hunk 1 above)
Comment (iycheng) [reply python/ray/_private/runtime_env.py:370]: Oh, not really. It has been moved to a separate file in `runtime_env`. Here it's for multiple envs, like the current env is too big, and we need to split them. In this case, we need to create a temp dir and link everything there. Right now, since we don't do that, we only have one uri.

File: src/ray/raylet/node_manager.cc (hunk 5 above)
Comment (iycheng) [reply src/ray/raylet/node_manager.cc:1248]: No, it's not. Previously, the file is "/path/hash.zip" now the directory is "/path/hash"

File: python/ray/_private/runtime_env.py (hunk 1 above)
Comment (edoakes) [reply python/ray/_private/runtime_env.py:371]: Gotcha, that makes sense.

**Review by edoakes — APPROVED** [review]

File: python/ray/worker.py [hunk 6]
```diff
@@ -1234,10 +1234,12 @@ def connect(node,
...
         # task spec and/or job config only.
-        job_config = os.environ.get("RAY_RUNTIME_ENV_FILES")
-        job_config = [job_config] if job_config else \
+        uris = os.environ.get("RAY_RUNTIME_ENV_FILES")
+        uris = [uris] if uris else \
             worker.core_worker.get_job_config().runtime_env.uris
-        runtime_env.ensure_runtime_env_setup(job_config)
+        working_dir = runtime_env.ensure_runtime_env_setup(uris)
+        if working_dir:
+            os.chdir(working_dir)
```
Comment (iycheng) [inline python/ray/worker.py:1242]: Here @edoakes

File: python/ray/util/client/server/server.py [hunk 7]
```diff
@@ -430,9 +430,9 @@ def _prepare_runtime_env(self, job_runtime_env) -> List[str]:
         with disable_client_hook():
             for uri in uris:
                 try:
-                    runtime_env.fetch_package(uri)
-                    print("Adding!: ", runtime_env._get_local_path(uri))
-                    sys.path.insert(0, str(runtime_env._get_local_path(uri)))
+                    pkg_dir = runtime_env.fetch_package(uri)
+                    if pkg_dir:
+                        os.chdir(pkg_dir)
```
Comment (iycheng) [inline python/ray/util/client/server/server.py:435]: Here @edoakes

File: python/ray/worker.py [hunk 8]
```diff
@@ -1276,16 +1278,19 @@ def connect(node,
...
+            # paths of the workers. Also add the current directory. Note that this
+            # assumes that the directory structures on the machines in the clusters
+            # are the same.
+            script_directory = os.path.abspath(os.path.dirname(sys.argv[0]))
+            current_directory = os.path.abspath(os.path.curdir)
+            worker.run_function_on_all_workers(
+                lambda worker_info: sys.path.insert(1, script_directory))
+            worker.run_function_on_all_workers(
+                lambda worker_info: sys.path.insert(1, current_directory))
         # TODO(rkn): Here we first export functions to run, then remote
```
Comment (iycheng) [inline python/ray/worker.py:1294]: @edoakes  could you take another look at this part? client-server will fail since the curdir is invalid (it got GCed once there is no instance using this). And for client-server, it's set up in another two places (please check my comment).

File: python/ray/worker.py (hunk 6 above)
Comment (edoakes) [reply python/ray/worker.py:1242]: LGTM

File: python/ray/worker.py [hunk 9]
```diff
@@ -1234,10 +1234,12 @@ def connect(node,
...
         # TODO(ekl) get rid of the env var hack and get runtime env from the
         # task spec and/or job config only.
-        job_config = os.environ.get("RAY_RUNTIME_ENV_FILES")
-        job_config = [job_config] if job_config else \
+        uris = os.environ.get("RAY_RUNTIME_ENV_FILES")
+        uris = [uris] if uris else \
             worker.core_worker.get_job_config().runtime_env.uris
-        runtime_env.ensure_runtime_env_setup(job_config)
+        working_dir = runtime_env.ensure_runtime_env_setup(uris)
+        if working_dir:
```
Comment (edoakes) [inline python/ray/worker.py:1241]:
```suggestion
        if working_dir is not None:
```
best practice

File: python/ray/util/client/server/server.py (hunk 7 above)
Comment (edoakes) [reply python/ray/util/client/server/server.py:435]: LGTM, any reason not to call this `working_dir` as well though?

Comment (iycheng) [reply python/ray/util/client/server/server.py:435]: sure
